crates/ruma-events/src/secret/request.rs

```diff
@@ -0,0 +1,381 @@
+//! Types for the *m.secret.request* event.
+
+use ruma_events_macros::EventContent;
+use ruma_identifiers::DeviceIdBox;
+use ruma_serde::StringEnum;
+use serde::{
+    de::{self, MapAccess, Visitor},
+    ser::SerializeMap,
+    Deserialize, Serialize,
+};
+
+use crate::ToDeviceEvent;
+
+/// Event sent by a client to request a secret from another device or to cancel a previous request.
+///
+/// It is sent as an unencrypted to-device event.
+pub type RequestToDeviceEvent = ToDeviceEvent<RequestToDeviceEventContent>;
+
+/// The payload for RequestToDeviceEvent.
+#[derive(Clone, Debug, Serialize, EventContent)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+#[ruma_event(type = "m.secret.request", kind = ToDevice)]
+pub struct RequestToDeviceEventContent {
+    /// The action for the request, one of `["request", "request_cancellation"]`.
+    ///
+    /// If the action is "request", the name of the secret must also be provided.
+    #[serde(flatten)]
+    pub action: RequestAction,
+
+    /// The ID of the device requesting the event.
+    pub requesting_device_id: DeviceIdBox,
+
+    /// A random string uniquely identifying (with respect to the requester and the target) the
+    /// target for a secret.
+    ///
+    /// If the secret is requested from multiple devices at the same time, the same ID may be used
+    /// for every target. The same ID is also used in order to cancel a previous request.
+    pub request_id: String,
+}
+
+impl RequestToDeviceEventContent {
+    /// Creates a new `RequestToDeviceEventContent` with the given action, requesting device ID and
+    /// request ID.
+    pub fn new(
+        action: RequestAction,
+        requesting_device_id: DeviceIdBox,
+        request_id: String,
+    ) -> Self {
+        Self { action, requesting_device_id, request_id }
+    }
+}
+
+impl<'de> Deserialize<'de> for RequestToDeviceEventContent {
```

Comment:
This impl can be derived; the logic for handling the action and name fields can instead be done in the `Deserialize` impl for `RequestAction` (which can also be implemented with a derive + `#[serde(try_from)]` so it's shorter than that visitor stuff).

Reply:
I'm not sure if I implemented it the way you meant, but it definitely is shorter and much less complex. Thanks for the pointers!